keps/sig-auth/1688-dynamic-authentication-config/README.md

```diff
@@ -0,0 +1,807 @@
+The API server supports a variety of command line flags that enable authentication
+via x509 certificates, OpenID Connect ID tokens, arbitrary bearer tokens via a
+token webhook, etc.  Dynamic authentication config expands this capability to a
+Kubernetes REST API.
+
+## Motivation
+
+<!--
+This section is for explicitly listing the motivation, goals and non-goals of
+this KEP.  Describe why the change is important and the benefits to users.  The
+motivation section can optionally provide links to [experience reports][] to
+demonstrate the interest in a KEP within the wider Kubernetes community.
+
+[experience reports]: https://github.com/golang/go/wiki/ExperienceReports
+-->
+
+Configuring authentication via command line flags has some limitations.  Some
+forms of config are simply easier to specify and understand in a structured
+manner such as via a Kubernetes API or file.  The existing authentication flags
+for OIDC and token webhook limit the user to only one of each of these types of
+authentication modes as there is no way to specify the set of flags multiple
+times [#71162].  These flag based configs also require an API server restart to
+take effect.
+
+Since there is no way to configure authentication other than via flags, hosted
+environments have no consistent way to expose this functionality [#166].  To get
+around this limitation, users install authentication proxies that run as
+cluster-admin and impersonate the desired user [kube-oidc-proxy] [teleport].
+Other than being a gross abuse of the impersonation API, this opens the API
+server to escalation bugs caused by the proxy (such as improper handling of
+incoming request headers).  This proxy also intercepts network traffic which
+poses privacy and scalability concerns.
+
+As a simple case study, it should be possible to deploy a simple identity
+provider (ex: [dex]) directly onto a running cluster and be able to use that IDP
+to authenticate users in the order of minutes without restarting the API server.
+No proxying of network traffic should occur and the impersonation API must not
+be used.
+
+[#71162]: https://github.com/kubernetes/kubernetes/issues/71162
+[#166]: https://github.com/aws/containers-roadmap/issues/166
+[kube-oidc-proxy]: https://aws.amazon.com/blogs/opensource/consistent-oidc-authentication-across-multiple-eks-clusters-using-kube-oidc-proxy
+[teleport]: https://aws.amazon.com/blogs/opensource/authenticating-eks-github-credentials-teleport
+[dex]: https://github.com/dexidp/dex
+
+### Goals
+
+<!--
+List the specific goals of the KEP.  What is it trying to achieve?  How will we
+know that this has succeeded?
+-->
+
+1. Create a new Kubernetes REST API that allows configuration of authentication
+  - x509
+  - OIDC
+  - Token Webhook
+2. Changes made via the REST API should be active in the order of minutes without
+requiring a restart of the API server
+3. Allow the use of a custom authentication stack in hosted Kubernetes offerings
```

Comment:
I’m not in agreement with this proposal and don’t agree that such a capability should be exposed as a built-in API within the community Kubernetes project. We would not enable this feature and we would desire that, if accepted, the feature is opt-in.

Reply:
Many many customers I talk to are looking to use k8s clusters from multiple providers and have consistent auth among them. This is something that comes up again and again with enterprise customers.

I think we should be looking at this from the point of view of the end users of the clusters (both cluster admins and app teams) and not hosting providers.  After all, they are the real drivers of the success of Kubernetes.  

@dcberg -- if this is something that users want and you don't enable it then you'll be at a competitive disadvantage.  That is how this should work.  Create the capability and then let the real users vote with their feet.

Reply:
Yes, Joe, thank you for putting it so well. In several organizations I've seen so far, the cloud provider's authentication system integrated into their Kubernetes offerings are of no use to us. They are merely an impediment, used only begrudgingly and thereafter ignored.

The cost is that these clusters wind up using very few principals with coarsely assigned permissions ("Fine, then every developer is a cluster administrator!"), for lack of finer-grained control that could be employed consistently across all the clusters hosted with different providers.

Audit logging is no longer accurate. RBAC is less useful. In all, the system becomes less secure and less trustworthy for important workload.

Reply:
@jbeda I don’t disagree that there is a need to have common auth across clusters. While such an approach would not break a hosted solution, it may limit the use of other integrations with the provider.
My ask is that the feature has an “opt-in” mechanism to allow providers flexibility to expose the feature in such as a way to limit impact and/or to point users to an alternative approach.